test/suites/scale/provisioning_test.go

```diff
@@ -0,0 +1,149 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scale_test
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	"github.com/samber/lo"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/test"
+	"github.com/aws/karpenter/pkg/apis/settings"
+	"github.com/aws/karpenter/pkg/apis/v1alpha1"
+	awstest "github.com/aws/karpenter/pkg/test"
+	"github.com/aws/karpenter/test/pkg/debug"
+	"github.com/aws/karpenter/test/pkg/environment/aws"
+)
+
+const testGroup = "provisioning"
+
+var _ = Describe("Provisioning", Label(debug.NoWatch), Label(debug.NoEvents), func() {
+	var provisioner *v1alpha5.Provisioner
+	var nodeTemplate *v1alpha1.AWSNodeTemplate
+	var deployment *appsv1.Deployment
+	var selector labels.Selector
+	var dsCount int
+
+	BeforeEach(func() {
+		nodeTemplate = awstest.AWSNodeTemplate(v1alpha1.AWSNodeTemplateSpec{AWS: v1alpha1.AWS{
+			SecurityGroupSelector: map[string]string{"karpenter.sh/discovery": settings.FromContext(env.Context).ClusterName},
+			SubnetSelector:        map[string]string{"karpenter.sh/discovery": settings.FromContext(env.Context).ClusterName},
+		}})
+		provisioner = test.Provisioner(test.ProvisionerOptions{
+			ProviderRef: &v1alpha5.MachineTemplateRef{
+				Name: nodeTemplate.Name,
+			},
+			Requirements: []v1.NodeSelectorRequirement{
+				{
+					Key:      v1alpha5.LabelCapacityType,
+					Operator: v1.NodeSelectorOpIn,
+					Values:   []string{v1alpha1.CapacityTypeOnDemand},
+				},
+				{
+					Key:      v1.LabelOSStable,
+					Operator: v1.NodeSelectorOpIn,
+					Values:   []string{string(v1.Linux)},
+				},
+			},
+			// No limits!!!
+			// https://tenor.com/view/chaos-gif-22919457
+			Limits: v1.ResourceList{},
+		})
+		deployment = test.Deployment(test.DeploymentOptions{
+			PodOptions: test.PodOptions{
+				ResourceRequirements: v1.ResourceRequirements{
+					Requests: v1.ResourceList{
+						v1.ResourceCPU:    resource.MustParse("100m"),
+						v1.ResourceMemory: resource.MustParse("100Mi"),
+					},
+				},
+				TerminationGracePeriodSeconds: lo.ToPtr[int64](0),
+			},
+		})
+		selector = labels.SelectorFromSet(deployment.Spec.Selector.MatchLabels)
+		// Get the DS pod count and use it to calculate the DS pod overhead
+		dsCount = env.GetDaemonSetCount(provisioner)
+	})
+	It("should scale successfully on a node-dense scale-up", func(_ context.Context) {
+		replicasPerNode := 1
+		expectedNodeCount := 500
+		replicas := replicasPerNode * expectedNodeCount
+
+		deployment.Spec.Replicas = lo.ToPtr[int32](int32(replicas))
+		// Hostname anti-affinity to require one pod on each node
```

Comment:
question: any reason you chose anti-affinity over choosing pod density? anti-affinity is definitely more expensive, so it'll be more of a strain on Karpenter, if that's what you want.
```
		provisioner.Spec.KubeletConfiguration = &v1alpha5.KubeletConfiguration{
			MaxPods: lo.ToPtr[int32](int32(maxPodDensity)),
		}
```

Reply:
Do you think it should be one over the other? I think anti-affinity makes sense, since it should be a higher bar for scheduling performance.